src/main/java/za/co/neslotech/spring/trainspringbootinsight/controller/CrudController.java

```diff
@@ -0,0 +1,102 @@
+
+/**
+ * A generic REST API Controller with CRUD operations.
+ * The child class should use the following annotations
+ *      1. @RestController
+ * @param <T> The entity class.
+ * @param <I> The primary key field type.
+ */
+public abstract class CrudController<T, I> {
+
+    protected abstract String[] getUpdateColumnExclusions();
+    protected abstract String getEntityId(T entity);
+    protected abstract T saveAndFlush(T entity);
+    protected abstract List<T> findAll();
+    protected abstract Optional<T> findById(I id);
+    protected abstract void delete(T entity);
+
+    @GetMapping
+    public ResponseEntity<?> list(){
+        return new ResponseEntity<>(findAll(), HttpStatus.OK);
+    }
+
+    @GetMapping
+    @RequestMapping("{id}")
+    public ResponseEntity<?> get(final @PathVariable I id){
+        try {
+            return ResponseEntity.ok(findById(id)
+                    .orElseThrow(() ->
+                            new EntityNotFoundException(String.format("Entity with id %s not found!", id))
+                    ));
+        }
+        catch (EntityNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
+        }
+    }
+
+    @PostMapping
+    public ResponseEntity<URI> create(final HttpServletRequest request, final @RequestBody T entity){
+        var newEntity = saveAndFlush(entity);
+        var location = URI.create(String.format("%s/%s", request.getRequestURL(),
+                getEntityId(newEntity)));
+        return ResponseEntity.created(location).body(location);
+    }
+
+    @RequestMapping(value = "{id}", method = RequestMethod.DELETE)
+    public ResponseEntity<String> deleteItem(final @PathVariable I id) {
+
+        try {
+            T entity = findById(id)
+                    .orElseThrow(() ->
+                            new EntityNotFoundException(String.format("Entity with id %s not found!", id))
+                    );
+
+            delete(entity);
+            return ResponseEntity.ok(String.format("Entity with id %s deleted successfully!", id));
+        }
+        catch (EntityNotFoundException e) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
+        }
+    }
+
+    @RequestMapping(value = "{id}", method = RequestMethod.PUT)
+    public ResponseEntity<?> update(final HttpServletRequest request, final @PathVariable I id,
+                                         final @RequestBody T entity){
+
+        try {
+            T existingEntity = findById(id)
+                    .orElseThrow(() ->
+                            new EntityNotFoundException(String.format("Entity with id %s not found!", id))
+                    );
+
+            BeanUtils.copyProperties(entity, existingEntity, getUpdateColumnExclusions());
+
+            var location = URI.create(request.getRequestURL().toString());
+            return ResponseEntity.status(HttpStatus.OK).location(location).body(location);
```

Comment:
I don't think we need a location for this endpoint, so lets update this

```suggestion
            return ResponseEntity.ok(existingEntity);
```